### PR TITLE
Select proper IP address for docker

### LINKED ISSUE #979: Docker Swarm Mode: traefik.docker.network should default only to networks that traefik is attached to
Assume the following situation:

* I'm using docker 1.12 in swarm mode
* overlay network `traefik-net` with 192.168.0.0/24
* overlay network `backend` with 192.168.1.0/24
* the traefik container is only attached to the `traefik-net` network with the ip address 192.168.0.2
* my application is attached to both networks because it needs to connect to a database container. Therefore it has two ip addresses: 192.168.0.4 and 192.168.1.2
* the application has the following labels set: `traefik.port=80` and `traefik.frontend.rule=myapp.example.com`

**Expected behavior:**
Traefik creates a backend for http://192.168.0.4

**Observed behavior in 1.1.2:**
Traefik alternates between http://192.168.0.4 and http://192.168.1.2 every time it refreshes its config which sporadically causes 502 Bad Gateway errors. This is clearly visible when refreshing the traefik web interface every 30 seconds or so.

**Possible solutions/fixes:**
* My current workaround: explicitly set `traefik.docker.network=traefik-net` for every single application that is attached to more than one network
* Simple solution: Allow setting a default value for `traefik.docker.network` in `traefik.toml` (as far as I can see from the docs this is not possible right now)
* The "it just works" solution: only consider IP addresses that lie within networks that traefik is attached to

### PR DESCRIPTION
**Pull request by tanyadegurechaff**
Fixes #979

### CODE REVIEW

File: provider/docker.go [hunk 1]
```diff
@@ -41,13 +42,14 @@ var _ Provider = (*Docker)(nil)
...
-	UseBindPortIP    bool       `description:"Use the ip address from the bound port, rather than from the inner network"`
-	SwarmMode        bool       `description:"Use Docker on Swarm Mode"`
+	BaseProvider       `mapstructure:",squash"`
+	Endpoint           string     `description:"Docker server endpoint. Can be a tcp or a unix socket endpoint"`
+	Domain             string     `description:"Default domain used"`
+	TLS                *ClientTLS `description:"Enable Docker TLS support"`
+	ExposedByDefault   bool       `description:"Expose containers by default"`
+	UseBindPortIP      bool       `description:"Use the ip address from the bound port, rather than from the inner network"`
+	SwarmMode          bool       `description:"Use Docker on Swarm Mode"`
+	traefikContainerID string
```
Comment (SantoDE) [inline provider/docker.go]: I'm not sure if that is a good idea or not. What happens, if you run traefik in cluster mode? How would you handle that case?

File: provider/docker.go [hunk 2]
```diff
@@ -128,6 +130,16 @@ func (provider *Docker) Provide(configurationChan chan<- types.ConfigMessage, po
...
 			version, err := dockerClient.ServerVersion(ctx)
 			log.Debugf("Docker connection established with docker %s (API %s)", version.Version, version.APIVersion)
+
+			hostname, err := os.Hostname()
+			if err != nil {
+				log.Errorf("Failed to get hostname for docker, error %s", err)
+				return err
+			}
+			if containerExists(ctx, dockerClient, hostname) {
+				provider.traefikContainerID = hostname
```
Comment (SantoDE) [inline provider/docker.go]: Same here. If you run it in cluster mode, wouldn't that be overwritten quite often and not beeing "stable"?

File: provider/docker.go [hunk 3]
```diff
@@ -719,3 +765,34 @@ func parseTasks(task swarmtypes.Task, serviceDockerData dockerData, networkMap m
...
 }
+
+func containerExists(ctx context.Context, cli client.ContainerAPIClient, id string) bool {
+	if _, err := cli.ContainerInspect(ctx, id); err != nil {
+		return false
+	}
+	return true
+}
+
+func getNetworks(ctx context.Context, cli client.APIClient, containerID string,
```
Comment (SantoDE) [inline provider/docker.go]: maybe something not that concrete, but we're having 2 ways to get the networks now (depending on the situation). I might need to think about that

**Comment by emilevauge** [comment]
Ping @tanyadegurechaff

**Comment by emilevauge** [comment]
Closing for inactivity.

**Pull request closed without merging**